### PR TITLE
Implement mainchain & sidechain recovery verification -  Closes #7022 

### LINKED ISSUE #7022: Implement mainchain Message Recovery Command verfication
### Description

Add message recovery command verification to mainchain interoperability module.

```
if terminatedOutboxAccount(trs.params.chainID) does not exist:
    return False

# Check that the CCMs are still pending
for index in trs.params.idxs:
    if index < terminatedOutboxAccount(trs.params.chainID).partnerChainInboxSize:
        return False

# Check the validity of the CCMs to be recovered
for ccm in deserializedCCMs:
    if ccm.status != CCM_STATUS_OK:
        return False

# Check the inclusion proof against the sidechain outbox root
proof = { size: terminatedOutboxAccount(trs.params.chainID).outboxSize,
    idxs: trs.params.idxs,
    siblingHashes: trs.params.siblingHashes}


return RMTVerify( 
  [SHA-256(ccmData) for ccmData in trs.params.crossChainMessages],
  proof,
  terminatedOutboxAccount(trs.params.chainID).outboxRoot
  )
```

### Acceptance Criteria 

- Add unit tests for message recovery

### PR DESCRIPTION
**Pull request by Incede**
### What was the problem?

This PR resolves #7022

### How was it solved?

Implement mainchain & sidechain recovery verification commands as specified

### How was it tested?

Added unit tests

### CODE REVIEW

File: framework/src/modules/interoperability/mainchain/commands/message_recovery.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import { CCMsg, StoreCallback, MessageRecoveryParams } from '../../types';
+import { BaseInteroperabilityCommand } from '../../base_interoperability_command';
+import { MainchainInteroperabilityStore } from '../store';
+import { getIDAsKeyForStore } from '../../utils';
+
+export class MessageRecoveryCommand extends BaseInteroperabilityCommand {
+	public id = COMMAND_ID_MESSAGE_RECOVERY;
+	public name = 'messageRecovery';
+
+	public async verify(
```
[the remark this answers is not in the document]
Comment (Incede) [reply framework/src/modules/interoperability/mainchain/commands/message_recovery.ts:35]: I am not sure since the stores(`this.getInteroperabilityStore(getStore)`)are not interchangeable but maybe you can elaborate how ?

[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/modules/interoperability/mainchain/commands/message_recovery.ts:35]: In @ishantiw 's PR https://github.com/LiskHQ/lisk-sdk/pull/7165, he unified the common logic into util functions

[the remark this answers is not in the document]
Comment (Incede) [reply framework/src/modules/interoperability/mainchain/commands/message_recovery.ts:31]: The common code is extracted(unified?) now.

File: framework/test/unit/modules/interoperability/sidechain/commands/message_recovery.spec.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+			{
+				outboxRoot,
+				outboxSize: terminatedChainOutboxSize,
+				partnerChainInboxSize: 1,
+			},
+			terminatedOutboxSchema,
+		);
+	});
+
+	it('should return error if the sidechain outbox root is not valid', async () => {
```
Comment (shuse2) [inline framework/test/unit/modules/interoperability/sidechain/commands/message_recovery.spec.ts:168]: Tests also seems to be duplicated. If the code is the same, do we need 2 same tests?

File: framework/src/modules/interoperability/sidechain/commands/message_recovery.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	public async verify(
+		context: CommandVerifyContext<MessageRecoveryParams>,
+	): Promise<VerificationResult> {
+		const {
+			params: { chainID, idxs, crossChainMessages, siblingHashes },
+			getStore,
+		} = context;
+		const chainIdAsBuffer = getIDAsKeyForStore(chainID);
+		const interoperabilityStore = this.getInteroperabilityStore(getStore);
+		let terminatedChainOutboxAccount: TerminatedOutboxAccount;
```
Comment (shuse2) [inline framework/src/modules/interoperability/sidechain/commands/message_recovery.ts:40]:
```suggestion
		let terminatedChainOutboxAccount: TerminatedOutboxAccount | undefined;
```
and you can remove the `verifyMessageRecovery` in the catch?

File: framework/test/unit/modules/interoperability/sidechain/commands/message_recovery.spec.ts (hunk 2 above)
Comment (Incede) [reply framework/test/unit/modules/interoperability/sidechain/commands/message_recovery.spec.ts]: Unit tests for `verify` hook under `Sidechain MessageRecoveryCommand` removed with a note.

**Review by mehmetegemen — CHANGES_REQUESTED** [review]

File: framework/src/modules/interoperability/base_interoperability_store.ts [hunk 4]
```diff
@@ -265,16 +282,18 @@ export abstract class BaseInteroperabilityStore {
 			STORE_PREFIX_TERMINATED_OUTBOX,
 		);
 
-		const terminatedOutboxAccount = await terminatedOutboxSubstore.getWithSchema<TerminatedOutbox>(
+		return terminatedOutboxSubstore.getWithSchema<TerminatedOutboxAccount>(
```
[the remark this answers is not in the document]
Comment (Incede) [reply framework/src/modules/interoperability/base_interoperability_store.ts:285]: It was changed due to merge conflicts. If you think it's wrong/not suitable feel free to suggest a better alternative.

File: framework/src/modules/interoperability/base_interoperability_store.ts [hunk 5]
```diff
@@ -89,7 +89,10 @@ export abstract class BaseInteroperabilityStore {
 	}
 
 	public async setOwnChainAccount(ownChainAccount: OwnChainAccount): Promise<void> {
-		const ownChainAccountStore = this.getStore(this.moduleID, STORE_PREFIX_OWN_CHAIN_DATA);
+		const ownChainAccountStore = this.getStore(
+			this.moduleID,
+			STORE_PREFIX_OWN_CHAIN_DATA,
+		) as SubStore;
```
[the remark this answers is not in the document]
Comment (Incede) [reply framework/src/modules/interoperability/base_interoperability_store.ts:95]: The verify hook of MessageRecoveryCommand needs access to interoperabilityStore but the return type of getStore in CommandVerifyContext is ImmutableSubStore. Hence the return type of getStore(StoreCallback) here is updatedto be SubStore | ImmutableStore. So, for functions other than getters the return type needs to be cast back to SubStore since ImmutableStore doesn’t support them.

File: framework/src/modules/interoperability/utils.ts [hunk 6]
```diff
@@ -107,6 +115,61 @@ export const computeValidatorsHash = (
...
+		size: terminatedChainOutboxAccount.outboxSize,
+		idxs,
+		siblingHashes,
+	};
+	const hashedCCMs = crossChainMessages.map(ccm => hash(ccm));
+	const isVerified = regularMerkleTree.verifyDataBlock(
+		hashedCCMs,
+		proof,
+		terminatedChainOutboxAccount.outboxRoot,
+	);
```
[the remark this answers is not in the document]
Comment (Incede) [reply framework/src/modules/interoperability/utils.ts:161]: I am not sure I understand what you mean. Maybe it's better if you suggest code blocks or exact logic because it's hard to follow otherwise.

**Review by mehmetegemen — APPROVED** [review]
